classes/robot/crawler.php

```diff
@@ -519,36 +519,77 @@ public function process_queue($verbose = false) {
         global $DB;
         $config = $this::get_config();
 
-        $recentcourses = false;
-        if ($config->uselogs == 1) {
-            $recentcourses = $this->get_recentcourses();
+        $allowedcourses = false;
+        if ($config->coursemode == 1) {
+            $allowedcourses = \tool_crawler\helper::get_onqueue_course_ids();
+        } else if ($config->uselogs == 1) {
+            $allowedcourses = $this->get_recentcourses();
         }
 
         // Iterate through the queue.
         $cronstart = time();
         $cronstop = $cronstart + $config->maxcrontime;
-        $hastime = true;
 
         // Get an instance of the currently configured lock_factory.
         $lockfactory = \core\lock\lock_config::get_lock_factory('tool_crawler_process_queue');
-
+        if ($config->coursemode == 1 || $config->uselogs == 1) {
+            if (empty($allowedcourses)) {
+                return true;
+            }
+            $courselockfactory = \core\lock\lock_config::get_lock_factory('tool_crawler_process_course_queue');
+            $courseid = 0;
+            foreach ($allowedcourses as $id) {
+                if ($courselock = $courselockfactory->get_lock($id, 0)) {
```

Comment:
why are you locking at the course level? You should only need to lock at the level of a specific url. Locking at the level will break the parallelism and be substantially slower and force the crawling to be in serial.

Reply:
The idea is to lock the crawling with in a course. It has to complete the process in a course before it can go to another one.
The parallelism should still be working if the url is on the current locked course.

Reply:
That doesn't sound like a feature that we'd actually want? If I am crawling one course then the system should be able to both crawl other courses at the same time, and crawl multiple urls within a course marked for crawl.

But my reading of this code means that you will only ever crawl a single url within a course at a time, EXCEPT that there is a bug in the code on line 549 where if you didn't get a lock then it ignores it and grabs the first course and then blindly keeps going when in theory it should have stopped. In general if you cannot get a lock for a certain thing then you must not keep going and do the thing. This means that the locking is broken for the first course in $allowedcourses.

I do not think there should be any locking at the course level, the only thing it will do is slow down the system overall.

Reply:
Regards "both crawl other courses at the same time, and crawl multiple urls within a course marked for crawl."

Currently we have configuration for number of parallel workers.
So what is the reasonable way to allocate the workers
ie, we have 10 workers and 5 queued courses.
Should we have 2 or 10 workers for each course? (Any limit if we make 10 for each course)
Or leave the crawling to run arbitrarily?

Reply:
I think in a first version it just be first in best dressed until you find problems to deal with and then solve them. If you only allow 1 course to run at a time then a course with 1000 urls to crawl which blocks another course with 10 urls from being crawled seems very unfair to me. A course with 1000 urls may also need multiple crawls to progressively discover all 1000 urls so it may not be as fast as simply 1000 / the number of tasks.

Also you cannot always know what course a given url is in until after you have crawled it the first time. I think there is an existing bug where the parsing for this is done in parse_html but this is never passed back and updated in tool_crawl_url so there will be urls missed here I think

Reply:
Hi @brendanheywood ,

I have removed the course lock. 

Unlike site crawling, on course mode, the crawler will need to choose a course to crawl (and I make it randomly) otherwise it will not know where to start.

Would you please have another review once you have time? Thanks